core/src/main/java/org/infinispan/commands/write/PutKeyValueCommand.java

```diff
@@ -96,13 +97,17 @@ public Object perform(InvocationContext ctx) throws Throwable {
       }
       MVCCEntry e = (MVCCEntry) ctx.lookupEntry(key);
 
-      //possible as in certain situations (e.g. when locking delegation is used) we don't wrap
-      if (e == null) return null;
+      if (e == null) {
+         throw new IllegalStateException("Not wrapped");
+      }
 
-      Object entryValue = e.getValue();
-      if (!valueMatcher.matches(e, null, value, valueEquivalence)) {
+      Object prevValue = e.getValue();
+      if (!valueMatcher.matches(prevValue, null, value, valueEquivalence)) {
          successful = false;
-         return entryValue;
+         return prevValue;
+      } else if (hasFlag(Flag.COMMAND_RETRY) && putIfAbsent) {
+         // TODO: hack branch!
+         e.setValue(null);
```

Comment:
It seems wrong for the listeners to get `previous value == null` when `putIfAbsent()` is retried and `previous value == new value` when a regular `put()` is retried in otherwise identical conditions.

Even if the previous value is now correct for `putIfAbsent()`, the listener can't assume all writes will be `putIfAbsent()`, so it should still check for `event.isCommandRetried()` and not trust the previous value when it's `true`. So I'm not sure this really helps the user or just adds another special case...